File: src/Model/PomniJson.cs
using System.Text.Json.Serialization;

namespace pomni.Model;

internal class PomniJson
{
    public required Forge Forge { get; init; }
    public required string Repository { get; init; }

    public string? Reference { get; init; }
    public ReferenceType? ReferenceType { get; init; }
    public bool? Frozen { get; init; }
}

internal enum Forge
{
    [JsonStringEnumMemberName("github")]
    GitHub,
}

internal enum ReferenceType
{
    [JsonStringEnumMemberName("branch")]
    Branch,

    [JsonStringEnumMemberName("release")]
    Release,
    
    [JsonStringEnumMemberName("tag")]
    Tag,
}

File: Sources/Pomni/Models/PomniJson.swift
import ArgumentParser

struct PomniJson: Codable {
  let version: Int
  var pins: [String: Pin] = [:]
}

struct Pin: Codable {
  var forge: Forge
  var repository: String
  var type: PinType?
  var branch: String?
  var frozen: Bool
}

enum Forge: String, Codable, ExpressibleByArgument {
  case forgejo, github
}

enum PinType: String, Codable, ExpressibleByArgument {
  case branch, release
}
